crates/red_knot_python_semantic/src/semantic_model.rs

```diff
@@ -147,24 +147,24 @@ impl HasTy for ast::Expr {
     }
 }
 
-macro_rules! impl_definition_has_ty {
+macro_rules! impl_binding_has_ty {
     ($ty: ty) => {
         impl HasTy for $ty {
             #[inline]
             fn ty<'db>(&self, model: &SemanticModel<'db>) -> Type<'db> {
                 let index = semantic_index(model.db, model.file);
-                let definition = index.definition(self);
-                definition_ty(model.db, definition)
+                let binding = index.definition(self);
+                binding_ty(model.db, binding)
             }
         }
     };
 }
 
-impl_definition_has_ty!(ast::StmtFunctionDef);
-impl_definition_has_ty!(ast::StmtClassDef);
-impl_definition_has_ty!(ast::Alias);
-impl_definition_has_ty!(ast::Parameter);
-impl_definition_has_ty!(ast::ParameterWithDefault);
+impl_binding_has_ty!(ast::StmtFunctionDef);
```

Comment:
This is interesting. What made you decide for using the inferred type over the declared type?

Reply:
Hmm, good question. It wasn't really a thought-through decision, I just didn't have any intent to make a change to this part of the code at all -- I was just updating the wording to clarify what it actually does.

I think the main reason to use inferred type here is that all of the definitions we currently implement this for always have an inferred type, and for the ones that have a declared type (functions, classes, parameters), it's always identical to the inferred type anyway. So using declared type here would never give a different result, except in the cases where it would give no result at all.

The only node which can have a different declared vs inferred type is AnnAssign, and we don't implement `HasTy` for it, partly for that very reason.